google-cloud-spanner/lib/google/cloud/spanner/pool.rb

```diff
@@ -154,32 +163,91 @@ def reset
 
         def close
           @thread.kill if @thread
-          @all_sessions.each(&:delete_session)
-          @all_sessions = []
-          @session_queue = []
-          @transaction_queue = []
+          # Unblock all waiting threads
+          @closed = true
+          @resource.broadcast
+          # Delete all sessions
+          @mutex.synchronize do
+            @all_sessions.map do |s|
+              Thread.new do
```

Comment:
It will be good to have some sort of bounds on how many threads we use. If you cannot use an external library for this, maybe you can do something like create 10 threads and have each thread close num_sessions/10 sessions sequentially in a loop.